File: sdk/python/coreauth/types/scim.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel


class ScimUser(BaseModel):
    schemas: Optional[list[str]] = None
    id: Optional[str] = None
    external_id: Optional[str] = None
    user_name: Optional[str] = None
    name: Optional[dict[str, str]] = None
    display_name: Optional[str] = None
    emails: Optional[list[dict[str, Any]]] = None
    phone_numbers: Optional[list[dict[str, Any]]] = None
    active: Optional[bool] = None
    groups: Optional[list[dict[str, Any]]] = None
    meta: Optional[dict[str, Any]] = None


class CreateScimUserRequest(BaseModel):
    schemas: Optional[list[str]] = None
    external_id: Optional[str] = None
    user_name: str
    name: dict[str, str]
    display_name: Optional[str] = None
    emails: list[dict[str, Any]] = []
    phone_numbers: Optional[list[dict[str, Any]]] = None
    active: bool = True
    password: Optional[str] = None


class ScimGroup(BaseModel):
    schemas: Optional[list[str]] = None
    id: Optional[str] = None
    external_id: Optional[str] = None
    display_name: Optional[str] = None
    members: Optional[list[dict[str, Any]]] = None
    meta: Optional[dict[str, Any]] = None


class CreateScimGroupRequest(BaseModel):
    schemas: Optional[list[str]] = None
    external_id: Optional[str] = None
    display_name: str
    members: Optional[list[dict[str, Any]]] = None


class ScimPatchRequest(BaseModel):
    schemas: Optional[list[str]] = None
    operations: list[dict[str, Any]] = []


class ScimListResponse(BaseModel):
    schemas: Optional[list[str]] = None
    total_results: int
    items_per_page: int
    start_index: int
    resources: list[dict[str, Any]] = []


class ScimTokenResponse(BaseModel):
    id: Optional[str] = None
    name: Optional[str] = None
    token_prefix: Optional[str] = None
    expires_at: Optional[str] = None
    created_at: Optional[str] = None


class ScimTokenWithSecret(BaseModel):
    id: Optional[str] = None
    name: Optional[str] = None
    token_prefix: Optional[str] = None
    expires_at: Optional[str] = None
    created_at: Optional[str] = None
    secret: Optional[str] = None


class CreateScimTokenRequest(BaseModel):
    name: str
    expires_at: Optional[str] = None


class SessionInfo(BaseModel):
    id: Optional[str] = None
    user_id: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    authenticated_at: Optional[str] = None
    last_active_at: Optional[str] = None
    expires_at: Optional[str] = None
    created_at: Optional[str] = None


class OidcProvider(BaseModel):
    id: Optional[str] = None
    tenant_id: Optional[str] = None
    name: Optional[str] = None
    provider_type: Optional[str] = None
    issuer: Optional[str] = None
    client_id: Optional[str] = None
    authorization_endpoint: Optional[str] = None
    token_endpoint: Optional[str] = None
    userinfo_endpoint: Optional[str] = None
    jwks_uri: Optional[str] = None
    scopes: Optional[list[str]] = None
    groups_claim: Optional[str] = None
    group_role_mappings: Optional[dict[str, str]] = None
    allowed_group_id: Optional[str] = None
    is_enabled: Optional[bool] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CreateOidcProviderRequest(BaseModel):
    tenant_id: str
    name: str
    provider_type: str
    issuer: str
    client_id: str
    client_secret: str
    authorization_endpoint: str
    token_endpoint: str
    userinfo_endpoint: Optional[str] = None
    jwks_uri: str
    scopes: Optional[list[str]] = None
    groups_claim: Optional[str] = None
    group_role_mappings: Optional[dict[str, str]] = None
    allowed_group_id: Optional[str] = None


class UpdateOidcProviderRequest(BaseModel):
    is_enabled: Optional[bool] = None
    name: Optional[str] = None
    client_id: Optional[str] = None
    client_secret: Optional[str] = None


class OidcLoginResponse(BaseModel):
    authorization_url: str
    state: str


class SsoCheckResponse(BaseModel):
    has_sso: bool
    providers: list[dict[str, Any]] = []

File: sdk/node/src/types/scim.ts
export interface ScimUser {
  schemas?: string[];
  id: string;
  externalId?: string;
  userName: string;
  name?: Record<string, any>;
  displayName?: string;
  emails?: Record<string, any>[];
  phoneNumbers?: Record<string, any>[];
  active?: boolean;
  groups?: Record<string, any>[];
  meta?: Record<string, any>;
}

export interface CreateScimUserRequest {
  schemas?: string[];
  externalId?: string;
  userName: string;
  name: Record<string, any>;
  displayName?: string;
  emails: Record<string, any>[];
  phoneNumbers?: Record<string, any>[];
  active?: boolean;
  password?: string;
}

export interface ScimGroup {
  schemas?: string[];
  id: string;
  externalId?: string;
  displayName: string;
  members?: Record<string, any>[];
  meta?: Record<string, any>;
}

export interface CreateScimGroupRequest {
  schemas?: string[];
  externalId?: string;
  displayName: string;
  members?: Record<string, any>[];
}

export interface ScimPatchRequest {
  schemas?: string[];
  Operations: { op: string; path?: string; value?: any }[];
}

export interface ScimListResponse {
  schemas?: string[];
  totalResults: number;
  itemsPerPage: number;
  startIndex: number;
  Resources: any[];
}

export interface ScimTokenResponse {
  id: string;
  name: string;
  token_prefix: string;
  expires_at?: string;
  created_at?: string;
}

export interface ScimTokenWithSecret extends ScimTokenResponse {
  secret: string;
}

export interface CreateScimTokenRequest {
  name: string;
  expires_at?: string;
}

export interface SessionInfo {
  id: string;
  user_id?: string;
  ip_address?: string;
  user_agent?: string;
  authenticated_at?: string;
  last_active_at?: string;
  expires_at?: string;
  created_at?: string;
}

export interface OidcProvider {
  id: string;
  tenant_id: string;
  name: string;
  provider_type: string;
  issuer: string;
  client_id: string;
  authorization_endpoint?: string;
  token_endpoint?: string;
  userinfo_endpoint?: string;
  jwks_uri?: string;
  scopes?: string[];
  groups_claim?: string;
  group_role_mappings?: Record<string, any>;
  allowed_group_id?: string;
  is_enabled: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface CreateOidcProviderRequest {
  tenant_id: string;
  name: string;
  provider_type: string;
  issuer: string;
  client_id: string;
  client_secret: string;
  authorization_endpoint: string;
  token_endpoint: string;
  userinfo_endpoint?: string;
  jwks_uri: string;
  scopes?: string[];
  groups_claim?: string;
  group_role_mappings?: Record<string, any>;
  allowed_group_id?: string;
}

export interface UpdateOidcProviderRequest {
  is_enabled?: boolean;
  name?: string;
  client_id?: string;
  client_secret?: string;
}

export interface SsoCheckResponse {
  has_sso: boolean;
  providers: Record<string, any>[];
}
